### DISCUSSION TITLE
[RFC] Metaprogramming and compiletime function and value evaluation/execution

### DISCUSSION THREAD

**Discussion by Janrupf** [modular/modular#218] [Ideas]
### What is this about

Mojo currently supports running some code when compiling, which results in values embedded in the final compilation result which have been pre-computed during the respective compilation. In fact, this is how Mojo's type system largely works and what at the time of writing this is called parameters.

This RFC is explicitly **not** about `@parameter` and `alias` - the discussion of unifying them is covered in #190.

Instead this RFC strives to define a powerful system for essentially running any code while compiling, how this can be achieved and how the syntax should look.

### Why is this necessary

Mojo's `@parameter` is somewhat powerful already and can be combined with `alias` perform extensive work upfront. However, in some places it still is severely limited, more specifically the following things are not working yet:

- Calling into C code 
- Generating closures at compiletime which then get transformed into raw functions (non-capturing)
-  (To be extended)

Calling into arbitrary code during compilation largely removes the need for complex build systems and external tooling. This also results in simpler and more easily reproducible builds as everything is self-contained.

### Where to start

For the sake of this RFC we probably want to introduce a new keyword or symbol - for the time being this can be `comptime`. The name is only a placeholder and subject to change as this RFC progresses. Further this intends to completely replace `@paramter` and `alias` (or redefine one of these).

### Which syntax should be used

Mojo currently uses `let` to bind immutable values in the form of `let x = 123`. This would open the possibility to either use a sigil or keyword as a prefix to bindings to make them statically evaluate during compilation (suggested by @nmsmith in #171).

This would mean using a sigil:
```mojo
~let x: Int = something()
~if cond:
    x = foo()
~else:
    x = bar()
```

Or using a keyword:

```mojo
comptime let x: Int = something()
comptime if cond:
    x = foo()
comptime else:
    x = bar()
```

Readability for both approaches seems to be about the same, however, the keyword makes it way more easy to directly spot. Additionally it offers (given the right choice) a more expressive and understandable indication of what is meant by it.
A sigil would have the advantage of being more concise, but seeing `~` (or any other sigil) doesn't directly convey the intention.

Should a keyword be used (chosen to temporarily be `comptime`) an in-depth discussion is required about the meaning. Mojo currently uses `parameter` to name anything executed/statically evaluated during compilation, which is in the context of other languages confusing. `comptime` may be equally confusing, as it doesn't convey what exactly is going on at compilation time. It however would have the advantage if being rather unambiguous with other languages. Naming certainly is an open ended discussion.

#### Execution

The arguably harder part is to define what is allowed in code executed during compilation (or just evaluated) and what not. The most ideal form would be a complete execution, where all Mojo code is valid and may run. This asks for a highly complex execution model but also provides the greatest flexibility.  To make it short: This would run Mojo while compiling Mojo, without restrictions.

Exactly opposite would be evaluation of simple expressions, mostly logic and math (including function calls).  While this significantly reduces flexibility, it also reduces complexity and minimizes possible problems.

#### Conditional logic

So far mainly 2 models have been proposed - statement/block and expression wise compiletime logic. These models are not exclusive of each other and both have their advantages and drawbacks.

Statement wise mostly affects control flow logic (this also intersects with syntax) and would applied to an `if` possibly look like this:
```mojo
comptime if expr:
    print("expr was true during compilation")
comptime elseif something:
    print("something was true during compilation")
else:
    print("none of the cases above were true during compilation")
```
It should be noted that only the conditionals are evaluated during compilation, the logic in the blocks continues to execute at runtime. The example also is missing a `comptime` in front of the `else`, as no condition is to be evaluated there.

This model would allow mixing compiletime and runtime conditions per branch:
```mojo
if a:
    print("Runtime a")
compiletime elseif b:
    print("Compiletime b")
elseif c:
    print("Runtime c")
else:
    print("None of the above")
```

However, a single `if` statement would with this model always be forced to only have compiletime or runtime conditions. A way around this would be to mark functions themselves as `comptime`. This however would severely restrict how they can be used and also suggest that a function has to be either run during compilation or during normal execution, but can't be both.

A more flexible approach is to mark single expression as run during compilation. Taking the example from above, this may look like this:
```mojo
if comptime(expr):
    print("expr was true during compilation")
elseif comptime(something):
    print("something was true during compilation")
else:
    print("none of the cases above were true during compilation")
```
The `if` continues to be an ordinary if, this possible is best demonstrated using `comptime` in another context:
```mojo
some_call(a, b, comptime(expr))
```
`a` and `b` would be regular variables, while `expr` would be run during compilation.

Going back to `if` (or generally control flow) statements this would then either rely on the optimizer to figure out a condition is constant during compilation or explicitly guarantee that `comptime` only control flow logic is always evaluated during compilation (this probably would happen either way due to optimization, but also provide a very explicit guarantee). 

Combining both approaches would also be viable, but may lead to somewhat weird constructs such as the following:
```mojo
comptime if comptime(expr):
    print("expr was true during compilation")
```
(`if comptime(expr)`, `comptime if expr` and `comptime if comptime(expr)` would all be equal).

#### Interleaving

Based on the 2 models from conditional logic, this can also be applied to blocks or singular expressions.
A compiletime block could look like the following:
```mojo
comptime:
    a()
    b()
    c()
```
There also have been suggestions of allowing deep nesting of `comptime` and `runtime`:
```mojo
comptime:
    a()
    runtime:
        b()
        comptime:
            c()
```
However, this seems super complex and could probably be achieved by simple re-ordering the calls (since compiletime code can never rely on runtime code for results anway).

Allowing `comptime` blocks also may lead to confusion regarding compiletime conditionals, such as the following example:
```mojo
comptime if expr:
    print("Is this run during compilation or execution?")
```
This is more of a "how could the syntax be improved?" question, but it ultimately leads to what to treat as a "block" for the sake of execution during compilation. 

Restricting `comptime` to expressions may look like this:
```mojo
comptime(a())
comptime(b())
comptime(c())
```

While a block would have been more concise, this example could be rewritten to have a singular function calling `a`, `b` and `c` and then having that been called by a `comptime` expression.

### Open questions
Everything, but specifically (in no particular order or completeness):

- [ ] How does this work with variable declarations?
- [ ] How can this replace `param_return` (if at all)
- [ ] How does this play with cross compilation
- [ ] Which problems arise
- [ ] Loops and their unrolling

I probably also missed some already brought up discussion points, they should be re-added to this RFC!

### Prior Art

- #190 
- #171 

### Addendum

All of the above is entirely open for discussion, some of it may also be incorrect or dreamed up for now. I hope that this can be used as a starting point to define how exactly this type of execution should look. This should be further edited and adjusted as the process goes on.

**Note: a lot of the work and ideas above are not mine but rather have been collected throughout the existing issues regarding this topic.**

**Comment by daydun** [comment]
I would like to stress, syntax (and especially syntax sugar) is not very relevant in the beginning. The interesting stuff is semantics. Syntax can be fixed up later once we have a solid grasp of how things should work.

Having a `comptime` per branch feels unnecessary and verbose.

The way it already works where `@parameter` evaluates all conditions at compile-time and only ever generates one (or zero if there's no catch-all else branch) branch for run-time execution is nice and predictable according to me, unless there's some clear example where it would be very useful to have per-branch control. If you need more fine-grained control you can always split it up into separate if-statements, where some are marked comp-time and some are runtime.

Being able to mark arbitrary expressions as compile-time with some inline syntax such as `comptime(x)`, `comptime[x]` or `comptime[x]()` could be useful. I believe this can sort of already be done today with
```py
fn comptime[t: AnyType, v: t]() -> t:
    return v

print(comptime[Int, 5]())
```

`comptime` blocks can already today be emulated by just running a `@parameter` function. Using a function also allows you to return stuff from the block (what else would the comp-time code be doing?).

Being able to reference runtime values inside of compile-time code sounds mildly intriguing. Would be interesting to try to come up with a realistic use case for that.

I have not used it very much myself, but if we want to fully embrace compile-time evaluation, the Zig language is probably a good source of inspiration as well. Not saying we should have a system identical to Zig, but exploring the problems they solve, and the issues they face is always good.

**Reply by nmsmith** [reply]
I don't think it's a good idea to assume that invoking a function using `[..]` brackets (your example: `comptime[Int, 5]()`) causes it to be executed at compile-time. As discussed [here](https://github.com/modularml/mojo/discussions/190#discussioncomment-5953424), it is theoretically possible—and highly desirable—to allow functions to be specialized (assigned parameters) at runtime.

**Reply by daydun** [reply]
That's a valid point. Though I would personally prefer if runtime specialization is done though some separate function / syntax such as `specialize[f]` or something, so that you don't accidentally ship a whole JIT engine on accident because a parameter was a run-time value.

**Reply by nmsmith** [reply]
That's true. I'm going to think about this for a while longer. I think we can come up with a very clear model for when code runs.

**Comment by daydun** [comment]
One thing I'd be interested in seeing is the ability to create new types using code at compile-time. For example stuff like
```py
fn main():
    @parameter
    fn f() -> AnyType:
        return Int
    
    alias x: f() = 5

main()
```

I believe Zig already supports this.

**Comment by nmsmith** [comment]
I'm tempted to make my own RFC now. I reckon we should avoid using the term "compile-time" while discussing Mojo's metaprogramming system. It's a bit of a mental trap. In general, specialization-time ≠ compile-time. [See here.](https://github.com/modularml/mojo/discussions/190#discussioncomment-5953424)

**Reply by daydun** [reply]
Good point. As far as I'm aware Zig does not really have any support for run-time specialization. I feel like there might be value in being able to clearly delimitate all three of compile-time, specialization-time (param-time? 😅) and run-time. Uncharted territory perhaps 🙂
